src/server/zset_family.cc

```diff
@@ -716,25 +731,38 @@ OpResult<ScoredMap> OpUnion(EngineShard* shard, Transaction* t, string_view dest
       continue;
 
     // first global index is 2 after {destkey, numkeys}
-    unsigned src_indx = j - start;
-    unsigned windex = t->ReverseArgIndex(shard->shard_id(), j) - 2;
-    DCHECK_LT(windex, weights.size());
-    it_arr[src_indx] = {*it_res, weights[windex]};
+    it_arr[j - start] = {*it_res, SliceResultWeight(shard, t, weights, j, offset)};
   }
+  return it_arr;
+}
 
-  ScoredMap result;
-  for (auto it = it_arr.begin(); it != it_arr.end(); ++it) {
-    if (it->first.is_done())
-      continue;
+OpResult<ScoredMap> OpUnion(EngineShard* shard, Transaction* t, string_view dest, AggType agg_type,
+                            const vector<double>& weights, bool store) {
+  ArgSlice keys = t->ShardArgsInShard(shard->shard_id());
+  DVLOG(1) << "shard:" << shard->shard_id() << ", keys " << vector(keys.begin(), keys.end());
+  DCHECK(!keys.empty());
 
-    ScoredMap sm = FromObject(it->first->second, it->second);
-    if (result.empty())
-      result.swap(sm);
-    else
-      UnionScoredMap(&result, &sm, agg_type);
+  unsigned start = 0;
+  unsigned offset = 0;
+
+  if (!dest.empty()) {
+    // first global index is 2 after {destkey, numkeys}
+    offset = 2;
+    if (keys.front() == dest) {
+      ++start;
+    }
+
+    // could be when only the dest key is hosted in this shard
```

Comment:
Please update comment
In case ONLY the destination key is hosted in this shard no work on this shard should be done in this step